### PR TITLE
feat(tooltip): display tooltip only when string is ellipsed

### PR DESCRIPTION
**Pull request by laraujo7**

### CODE REVIEW

File: packages/ui-components/src/components/info-label/info-label.tsx [hunk 1]
```diff
@@ -105,7 +105,7 @@ export class KvInfoLabel {
 							<div class="description">
 								{this.description && <div class="text">{this.description}</div>}
 								{this.copyValue && (
-									<kv-tooltip text={this.tooltipConfig.label} position={this.tooltipConfig.position}>
+									<kv-tooltip text={this.tooltipConfig.label} position={this.tooltipConfig.position} display={true}>
```
Comment (nelsonmestevao) [inline packages/ui-components/src/components/info-label/info-label.tsx:108]:
```suggestion
									<kv-tooltip text={this.tooltipConfig.label} position={this.tooltipConfig.position} display>
```

Is this valid?

Comment (laraujo7) [reply packages/ui-components/src/components/info-label/info-label.tsx:108]: Yap 😅

File: packages/ui-components/src/components/text-field/test/text-field.e2e.ts [hunk 2]
```diff
@@ -1,212 +0,0 @@
-import { E2EPage, newE2EPage, EventSpy, E2EElement } from '@stencil/core/testing';
```
Comment (djcouto) [inline packages/ui-components/src/components/text-field/test/text-field.e2e.ts:1]: Why did you remove this?

File: packages/ui-components/src/components/text-field/test/text-field.spec.tsx [hunk 3]
```diff
@@ -1,46 +0,0 @@
-import { SpecPage } from '@stencil/core/internal';
-import { KvTextField } from '../text-field';
-import { newSpecPage } from '@stencil/core/testing';
-
-describe('Text Field (unit tests)', () => {
```
Comment (djcouto) [inline packages/ui-components/src/components/text-field/test/text-field.spec.tsx:5]: And also this?

File: packages/ui-components/src/components/tooltip/tooltip.tsx [hunk 4]
```diff
@@ -28,6 +28,8 @@ export class KvTooltip implements ITooltip {
 	@Prop({ reflect: true }) disabled?: boolean = false;
 	/** @inheritdoc */
 	@Prop({ reflect: false }) contentElement?: HTMLElement = null;
+	/** @inheritdoc */
+	@Prop({ reflect: false }) display?: boolean = false;
```
Comment (djcouto) [inline packages/ui-components/src/components/tooltip/tooltip.tsx:32]: Despite this being a good way to handle it, I have a different suggestion, which IMO is cleaner. I think you should have a prop called 'behaviour' that would be either 'click', 'hover' or 'truncate'. Then, the tooltip will only be shown if either one of the following conditions is met:
- Behaviour is 'click' and the tooltip content is clicked.
- Behaviour is 'hover' and the tooltip content is with the mouse over.
- Behaviour is 'truncate', the tooltip content width is not enough and is truncated and it is with the mouse over.

What do you think?

**Review by djcouto — CHANGES_REQUESTED** [review]

File: packages/ui-components/src/components/tooltip/tooltip.tsx (hunk 4 above)
Comment (laraujo7) [reply packages/ui-components/src/components/tooltip/tooltip.tsx:32]: Yeap, I didn't particularly appreciate how I did this. That's an excellent suggestion I like it a lot  💪

File: packages/ui-components/src/components/tooltip/tooltip.tsx [hunk 5]
```diff
@@ -55,12 +57,23 @@ export class KvTooltip implements ITooltip {
...
-		if (this.delay) {
-			this.timeoutDelayId = window.setTimeout(() => {
-				this.showTooltip();
-			}, this.delay);
-		} else {
-			this.showTooltip();
+		const range = new Range();
+		range.selectNodeContents(this.el);
+
+		switch (this.behaviour) {
```
Comment (miguelpinto98) [inline packages/ui-components/src/components/tooltip/tooltip.tsx:63]: Since all the cases can use the same resource code, I suggest to change this in a more readable way:

```
if (this.behaviour === ETooltipBehaviour.Truncate) {
	const { width: rectWidth } = range.getBoundingClientRect();
	const { width: elWidth } = this.el.getBoundingClientRect();

	if (rectWidth < elWidth) return;
}

if (this.delay) {
	this.timeoutDelayId = window.setTimeout(() => {
		this.showTooltip();
	}, this.delay);
} else {
	this.showTooltip();
}
```

File: packages/ui-components/src/components/select/readme.md [hunk 6]
```diff
@@ -119,3 +119,5 @@ graph TD;
 ```
 
 ----------------------------------------------
+
```
Comment (miguelpinto98) [inline packages/ui-components/src/components/select/readme.md:122]: remove changes?

File: packages/ui-components/src/components/tooltip/tooltip.tsx (hunk 4 above)
Comment (miguelpinto98) [reply packages/ui-components/src/components/tooltip/tooltip.tsx]: I have some doubts about click behaviour. What we intent with it? Only open the tooltip if a click is done in the element? Or it's only to have the cursor point in the element?

**Review by miguelpinto98 — CHANGES_REQUESTED** [review]
Awesome improvement, but I left some suggestions. I think we also need to express this changes in our storybook, so please change the tooltip story with this new behaviours. To this PR, also add some description of the problem.

File: packages/ui-components/src/components/tooltip/tooltip.tsx (hunk 4 above)
Comment (ricardogoncalves89) [reply packages/ui-components/src/components/tooltip/tooltip.tsx]: Be careful, we already have a tooltip for click behavior with name kv-toggle-tip. Please speak with @GoncaloCamaz before startiing.

**Review by djcouto — CHANGES_REQUESTED** [review]
LGMT 👍 

Please just add stories with each tooltip behaviour.

**Comment by laraujo7** [comment]
@miguelpinto98 @ricardogoncalves89 @djcouto I change a little bit the strategy, and right now,  I'm searching for the first HTMLElement that has the ellipsis style, and after that, I'm calculating the difference between the Element found and the inner Element. Let me know what you think :muscle:

File: packages/ui-components/src/components/tooltip/tooltip.types.ts [hunk 7]
```diff
@@ -12,4 +12,11 @@ export interface ITooltip {
 	disabled?: boolean;
 	/** (optional) Content element reference */
 	contentElement?: HTMLElement;
+	/** (optional) Display tooltip */
+	behaviour?: ETooltipBehaviour;
+}
+
+export enum ETooltipBehaviour {
+	Hover = 'hover',
```
Comment (djcouto) [inline packages/ui-components/src/components/tooltip/tooltip.types.ts:20]: Sorry, but this seems a bit odd to me. Since we are not adding a "click" behaviour to the list, seems that the 'hover' and 'truncate' are the same behaviour. The difference is that we use 'truncate' the child container will be ellipsed/truncated.

Can you refactor this so the truncate becomes a boolean prop?

Sorry for the trouble.

File: packages/ui-components/src/components/utils.ts [hunk 8]
```diff
@@ -12,3 +12,19 @@ export const getConfig = (): UIComponentsConfig => {
 
 	return DEFAULT_CONFIG;
 };
+
+export const getCollapsedElement = (elements: NodeListOf<ChildNode>): HTMLElement => {
```
Comment (djcouto) [inline packages/ui-components/src/components/utils.ts:16]: This function signature must be:

```suggestion
export const getCollapsedElement = (elements: NodeListOf<ChildNode>): HTMLElement | undefined => {
```

Correct?

**Review by djcouto — CHANGES_REQUESTED** [review]
LGTM! 👍 

However, I want to you try to change a little thing. It should be quick, nevertheless, sorry for your trouble. 🥲

File: packages/ui-components/src/components/utils.ts (hunk 8 above)
Comment (miguelpinto98) [inline packages/ui-components/src/components/utils.ts:16]: what you think in clear a little bit this function?

my suggestion:
```
export const getCollapsedElement = (elements: NodeListOf<ChildNode>): HTMLElement | undefined => {
	if (elements.length === 0) return;

	return Array.from(elements).find(node => {
		if (getCssStyle(node as HTMLElement, 'text-overflow') === 'ellipsis') {
			return node;
		}

		return getCollapsedElement(node.childNodes);
	}) as HTMLElement;
};
```

**Review by miguelpinto98 — CHANGES_REQUESTED** [review]
Great improvement too! 🚀  Almost done, only small suggestions to base code.
Other thing, put this code in a alpha version to test with a real use case before the merge.

File: packages/ui-components/src/components/utils.ts (hunk 8 above)
Comment (laraujo7) [reply packages/ui-components/src/components/utils.ts]: Won't this return the first parent of the Element found?

File: packages/ui-components/src/components/tooltip/tooltip.tsx [hunk 9]
```diff
@@ -55,6 +58,19 @@ export class KvTooltip implements ITooltip {
 	};
 
 	private showTooltipHandler = () => {
+		if (this.truncate) {
+			const collapsedElement = getCollapsedElement(this.el.childNodes);
```
Comment (djcouto) [inline packages/ui-components/src/components/tooltip/tooltip.tsx:62]: Maybe it would be a nice thing to move this logic to a helper. Something like, `isElementCollpased(element: HTMLElement): boolean`?

File: packages/ui-components/src/components/tree-item/readme.md [hunk 10]
```diff
@@ -135,42 +135,6 @@ export const TreeItemExample: React.FC = () => (
 | `"children"` | The children container. |
 
 
-## CSS Custom Properties
```
Comment (djcouto) [inline packages/ui-components/src/components/tree-item/readme.md:138]: Revert this 🙏

File: packages/ui-components/src/components/tooltip/readme.md [hunk 11]
```diff
@@ -75,13 +76,6 @@ export const TagLetterExample: React.FC = () => (
 | `"content"`   | The tooltip content.   |
 
 
-## CSS Custom Properties
-
-| Name                  | Description                                  |
```
Comment (djcouto) [inline packages/ui-components/src/components/tooltip/readme.md:80]: Revert this 🙏

File: packages/ui-components/src/components/utils.ts [hunk 12]
```diff
@@ -12,3 +12,19 @@ export const getConfig = (): UIComponentsConfig => {
 
 	return DEFAULT_CONFIG;
 };
+
+export const getCollapsedElement = (elements: NodeListOf<ChildNode>): HTMLElement | undefined => {
```
Comment (djcouto) [inline packages/ui-components/src/components/utils.ts:16]: I know, everyone has a suggestion 😄 

Here is mine:

```ts
export const getCollapsedElement = (elements: NodeListOf<ChildNode>): HTMLElement | undefined => {
	if (elements.length === 0) {
		return;
	}

	const [firstElement, ...siblings] = elements;

	if (isOverflowEllipsis(firstElement as HTMLElement)) {
		return firstElement;
	}

	return getCollapsedElement(siblings);
}

export const isOverflowEllipsis = (element: HTMLElement): boolean =>
	isNodeText(element) && getCssStyle(element, "text-overflow") === "ellipsis";

export const isNodeText = (element: HTMLElement): boolean =>
  	element.nodeName !== "#text";

export const getCssStyle = (element: HTMLElement, prop: string): string =>
  	window.getComputedStyle(element, null).getPropertyValue(prop);

```

**Review by djcouto — CHANGES_REQUESTED** [review]
Final comments, great work 👍

**Review by djcouto — APPROVED** [review]
🚀

File: packages/ui-components/src/components/tooltip/tooltip.types.ts [hunk 13]
```diff
@@ -12,4 +12,6 @@ export interface ITooltip {
 	disabled?: boolean;
 	/** (optional) Content element reference */
 	contentElement?: HTMLElement;
+	/** (optional) Display tooltip */
```
Comment (ricardogoncalves89) [inline packages/ui-components/src/components/tooltip/tooltip.types.ts:15]: Add a better description please.

File: packages/ui-components/src/components/utils.ts [hunk 14]
```diff
@@ -12,3 +12,23 @@ export const getConfig = (): UIComponentsConfig => {
...
+	if (isOverflowEllipsis(firstElement as HTMLElement)) {
+		return firstElement;
+	}
+
+	return getCollapsedElement(Array.from(firstElement.childNodes)) ?? getCollapsedElement(siblings);
+};
+
+export const isOverflowEllipsis = (element: HTMLElement): boolean => isNodeText(element) && getCssStyle(element, 'text-overflow') === 'ellipsis';
+
+export const isNodeText = (element: HTMLElement): boolean => element.nodeName !== '#text';
```
Comment (ricardogoncalves89) [inline packages/ui-components/src/components/utils.ts:32]: Node is a Text node if nodeName is different from `#text`??

**Review by ricardogoncalves89 — APPROVED** [review]
LGTM. check the suggestions please

File: packages/ui-components/src/components/utils.ts (hunk 14 above)
Comment (laraujo7) [reply packages/ui-components/src/components/utils.ts:32]: Nope, that's wrong. I'm going to change it, nice catch :muscle:

**Review by miguelpinto98 — APPROVED** [review]
🚀
